File: source/m-files/arcCorrection.m
function [x, y, options] = arcCorrection(options, interpolateSinogram)
%ARCCORRECTION Performs arc correction on the detector coordinates
%   This function outputs the arc corrected detector coordinates and
%   sinogram. Works only with sinogram data and without precomputation.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Copyright (C) 2020-2025 Ville-Veikko Wettenhovi
%
% This program is free software: you can redistribute it and/or modify it
% under the terms of the GNU General Public License as published by the
% Free Software Foundation, either version 3 of the License, or (at your
% option) any later version.
%
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
% Public License for more details.
%
% You should have received a copy of the GNU General Public License along
% with this program. If not, see <https://www.gnu.org/licenses/>.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% mashing = options.det_w_pseudo / options.Nang / 2;
[~, ~, xp, yp] = detector_coordinates(options);
orig_xp = xp;
orig_yp = yp;
[x_o, y_o] = sinogram_coordinates_2D(options, orig_xp, orig_yp);

new_xp = zeros(size(xp));
new_yp = zeros(size(yp));
% The angles of the blocks/buckets
if mod(options.blocks_per_ring, 4) == 0
    l_angles = linspace(0,90, options.blocks_per_ring / 4 + 1);
else
    l_angles = linspace(0,180, options.blocks_per_ring / 2 + 1);
    l_angles = l_angles(1 : ceil(options.blocks_per_ring / 4));
end
ll = 1;

% Shift to the zero angle (with x-axis)
if options.det_w_pseudo > options.det_per_ring
    xp = circshift(xp, -options.offangle - (options.cryst_per_block + 1) / 2);
    yp = circshift(yp, -options.offangle - (options.cryst_per_block + 1) / 2);
else
    xp = circshift(xp, -options.offangle - options.cryst_per_block / 2);
    yp = circshift(yp, -options.offangle - options.cryst_per_block / 2);
end

xp = xp - options.diameter / 2;
yp = yp - options.diameter / 2;

% Determine the points along the circle that reside (approximately) on the
% same line as the original LOR
for kk = 1 : length(xp) / 4
    if options.det_w_pseudo > options.det_per_ring
        vali = kk + options.det_w_pseudo / 2 - (options.cryst_per_block + 1) : kk + options.det_w_pseudo / 2 + (options.cryst_per_block + 1);
    else
        vali = kk + options.det_w_pseudo / 2 - options.cryst_per_block : kk + options.det_w_pseudo / 2 + options.cryst_per_block;
    end
    kulma = l_angles(ll);
    angle1 = atand((yp(kk)-yp(vali))./(xp(kk)-xp(vali)));
    angle1 = round(angle1*10^4)/10^4;
    if mod(options.blocks_per_ring, 4) == 0
        ind1 = find(abs(angle1) == kulma);
        if isempty(ind1)
            [~, ind1] = min(abs(abs(angle1) - kulma));
        end
    else
        [~, ind1] = min(abs(abs(angle1) - kulma));
    end
    if options.det_w_pseudo > options.det_per_ring
        x2 = xp(kk + options.det_w_pseudo / 2 - (options.cryst_per_block + 1) + ind1 - 1);
        y2 = yp(kk + options.det_w_pseudo / 2 - (options.cryst_per_block + 1) + ind1 - 1);
    else
        x2 = xp(kk + options.det_w_pseudo / 2 - options.cryst_per_block + ind1 - 1);
        y2 = yp(kk + options.det_w_pseudo / 2 - options.cryst_per_block + ind1 - 1);
    end
    p = [xp(kk); yp(kk)];
    q = [x2; y2];
    d = q - p;
    l = ((-dot(2*p,d) - sqrt(dot(2*p,d)^2 - 4*dot(d,d)*(dot(p,p) - (options.diameter/2)^2)))/ (2*dot(d,d)));
    lx = p + l * d;
    new_xp(kk) = lx(1);
    new_yp(kk) = lx(2);
end

new_xp(1 : length(xp) / 4) = new_xp(1 : length(xp) / 4) + options.diameter/2;
new_yp(1 : length(yp) / 4) = new_yp(1 : length(yp) / 4) + options.diameter/2;

new_yp(kk + 1: kk * 2) = flip(new_yp(1: kk));
diffi = diff([flip(new_yp(1 : kk)) ; options.diameter/2]);
diffi = options.diameter/2 + cumsum(flip(diffi));
new_yp(kk * 2 + 1 : end) = [diffi ; flip(diffi)];

diffi = diff([new_xp(1 : kk) ; options.diameter/2]);
diffi = options.diameter/2 + cumsum(flip(diffi));
new_xp(kk + 1: kk * 2) = diffi;
new_xp(kk * 2 + 1 : end) = flip(new_xp(1 : kk * 2));
xp = new_xp;
yp = new_yp;

if options.det_w_pseudo > options.det_per_ring
    xp = circshift(xp, options.offangle + (options.cryst_per_block + 1) / 2);
    yp = circshift(yp, options.offangle + (options.cryst_per_block + 1) / 2);
else
    xp = circshift(xp, options.offangle + options.cryst_per_block / 2);
    yp = circshift(yp, options.offangle + options.cryst_per_block / 2);
end

[x, y] = sinogram_coordinates_2D(options, xp, yp);


xx1 = reshape(x(:,1),options.Ndist,options.Nang);
xx2 = reshape(x(:,2),options.Ndist,options.Nang);
yy1 = reshape(y(:,1),options.Ndist,options.Nang);
yy2 = reshape(y(:,2),options.Ndist,options.Nang);

% y_orig = y;
% x_orig = x;

angle = atand((yy1-yy2)./(xx1-xx2)) + 90;
angle(angle == 180) = 0;
[~, J] = min(mean(angle));

% Create the arc corrected coordinates from the perpendicular LOR
% coordinates
eka = xx1(1,J);
vika = xx1(end,J);
alkux = linspace(eka, vika, options.Ndist);

% Determine the y-coordinates
alkuy = sqrt((options.diameter/2)^2 - (alkux - options.diameter/2).^2) + options.diameter/2;

alku = [alkux; alkuy];

alku2 = alku - options.diameter/2;

alku = [alkux; abs(alkuy - options.diameter)];

alku1 = alku - options.diameter/2;


angles = linspace(0, 180, options.Nang + 1);
angles = angles(1 : end - 1);
angles = circshift(angles, J);
angles = reshape(angles, 1, 1, []);

% Rotation matrix
rot_matrix = [cosd(angles) -sind(angles); sind(angles) cosd(angles)];

rot_matrix = squeeze(num2cell(rot_matrix, [1 2]))';

% Arc correction
% Rotate the arc corrected coordinates for the whole ring
new_xy1 = cell2mat(cellfun(@(x) x * alku1, rot_matrix, 'UniformOutput', false))' + options.diameter/2;
new_xy2 = cell2mat(cellfun(@(x) x * alku2, rot_matrix, 'UniformOutput', false))' + options.diameter/2;


x = [new_xy1(:,1), new_xy2(:,1)];
y = [new_xy1(:,2), new_xy2(:,2)];

xx1 = reshape(x(:,1),options.Ndist,options.Nang);
xx2 = reshape(x(:,2),options.Ndist,options.Nang);
apu = xx1(:, 1: J);
xx1(:, 1: J) = flipud(xx2(:, 1: J));
xx2(:, 1: J) = flipud(apu);
yy1 = reshape(y(:,1),options.Ndist,options.Nang);
yy2 = reshape(y(:,2),options.Ndist,options.Nang);
apu = yy1(:, 1: J);
yy1(:, 1: J) = flipud(yy2(:, 1: J));
yy2(:, 1: J) = flipud(apu);

x(:,1) = xx1(:);
x(:,2) = xx2(:);
y(:,1) = yy1(:);
y(:,2) = yy2(:);

xx1_o = reshape(x_o(:,1),options.Ndist,options.Nang);
xx2_o = reshape(x_o(:,2),options.Ndist,options.Nang);
yy1_o = reshape(y_o(:,1),options.Ndist,options.Nang);
yy2_o = reshape(y_o(:,2),options.Ndist,options.Nang);

x = x - options.diameter / 2;
y = y - options.diameter / 2;

if interpolateSinogram
    
    
    % Angles
    angle_o = atand((yy1_o-yy2_o)./(xx1_o-xx2_o));
    angle_o = angle_o - min(angle_o(:,1));
    angle_o(angle_o<0) = angle_o(angle_o<0) + 180;
    angle_o = [-angle_o(:,2), angle_o, abs(angle_o(:,1)-180)];
    angle = atand((yy1-yy2)./(xx1-xx2));
    angle = angle - min(angle(:,1));
    angle(angle<0) = angle(angle<0) + 180;
    angle = [-angle(:,2), angle, abs(angle(:,1)-180)];
    
    
    % Orthogonal distances
    % x0 = options.diameter/2;
    % y0 = options.diameter/2;
    x0 = 0;
    y0 = 0;
    distance_o = ((abs((y_o(:,1)-y_o(:,2))*x0 - (x_o(:,1) - x_o(:,2))*y0 + x_o(:,1).*y_o(:,2) - y_o(:,1).*x_o(:,2))./sqrt((y_o(:,1)-y_o(:,2)).^2 + (x_o(:,1)-x_o(:,2)).^2)));
    distance_o = reshape(distance_o, options.Ndist, options.Nang);
    distance_o(options.Ndist/2 + 1 : end,:) = -distance_o(options.Ndist/2 + 1 : end,:);
    distance_o = [distance_o(:,1), distance_o, distance_o(:,1)];
    distance = ((abs((y(:,1)-y(:,2))*x0 - (x(:,1) - x(:,2))*y0 + x(:,1).*y(:,2) - y(:,1).*x(:,2))./sqrt((y(:,1)-y(:,2)).^2 + (x(:,1)-x(:,2)).^2)));
    distance = reshape(distance, options.Ndist, options.Nang);
    distance(options.Ndist/2 + 1 : end,:) = -distance(options.Ndist/2 + 1 : end,:);
    distance = [distance(:,1), distance, distance(:,1)];
    testi = diff(distance);
    if round(min(testi(:)),3) ~= round(max(testi(:)),3)
        warning('Arc correction failed to make all the LORs equidistant.')
    end
    
    
    % Interpolate the sinogram
    if iscell(options.SinM)
        if numel(options.SinM{1}) == size(options.SinM{1},1)
            for kk = 1 : options.partitions
                options.SinM{kk} = reshape(options.SinM{kk}, options.Ndist, options.Nang, numel(options.SinM)/(options.Ndist * options.Nang));
            end
        end
    else
        if numel(options.SinM) == size(options.SinM,1) && ~isscalar(options.SinM)
            options.SinM = reshape(options.SinM, options.Ndist, options.Nang, options.NSinos, []);
        end
    end
    tic
    if iscell(options.SinM)
        uus_SinM = cell(size(options.SinM));
        for hh = 1 : options.partitions
            for uu = 1 : size(apu_SinM,4)
                apu_SinM = zeros(size(options.SinM{hh},1),size(options.SinM{hh},2)+2,size(options.SinM{hh},3));
                % Use parfor if available
                if license('test','Distrib_Computing_Toolbox')
                    temp = options.SinM{hh}(:,:,:,uu);
                    temp = [temp(:,end,:), temp, temp(:,1,:)];
                    Ndist = options.Ndist;
                    Nang = options.Nang;
                    arc_interpolation = options.arc_interpolation;
                    try
                        parfor kk = 1 : size(options.SinM,3)
                            if exist('scatteredInterpolant', 'file') == 2 && (~strcmp('cubic',arc_interpolation) && ~strcmp('v4',arc_interpolation))
                                F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double(temp(:,:,kk)), [],1));
                                apu_SinM(:,:,kk) = F(angle, distance);
                            else
                                apu_SinM(:,:,kk) = griddata(angle_o, distance_o, double(temp(:,:,kk)), angle, distance, arc_interpolation);
                            end
                        end
                    catch
                        for kk = 1 : size(options.SinM,3)
                            if exist('scatteredInterpolant', 'file') == 2 && (~strcmp('cubic',arc_interpolation) && ~strcmp('v4',arc_interpolation))
                                F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double([options.SinM{hh}(:,end,kk,uu), options.SinM{hh}(:,:,kk,uu), options.SinM{hh}(:,1,kk,uu)]), [],1));
                                apu_SinM(:,:,kk) = F(angle, distance);
                            else
                                apu_SinM(:,:,kk) = griddata(angle_o, distance_o, double([options.SinM{hh}(:,end,kk,uu), options.SinM{hh}(:,:,kk,uu), options.SinM{hh}(:,1,kk,uu)]), angle, distance, arc_interpolation);
                            end
                        end
                    end
                else
                    for kk = 1 : size(options.SinM,3)
                        if exist('scatteredInterpolant', 'file') == 2 && (~strcmp('cubic',options.arc_interpolation) && ~strcmp('v4',options.arc_interpolation))
                            F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double([options.SinM{hh}(:,end,kk,uu), options.SinM{hh}(:,:,kk,uu), options.SinM{hh}(:,1,kk,uu)]), [],1));
                            apu_SinM(:,:,kk) = F(angle, distance);
                        else
                            apu_SinM(:,:,kk) = griddata(angle_o, distance_o, double([options.SinM{hh}(:,end,kk,uu), options.SinM{hh}(:,:,kk,uu), options.SinM{hh}(:,1,kk,uu)]), angle, distance, options.arc_interpolation);
                        end
                    end
                end
                apu_SinM(isnan(apu_SinM)) = single(0);
                uus_SinM{hh}(:,:,:,uu) = single(apu_SinM);
            end
            endTime = toc;
            options.SinM{hh} = uus_SinM{hh};
        end
    else
        for uu = 1 : size(options.SinM,4)
            uus_SinM = zeros(size(options.SinM,1),size(options.SinM,2)+2,size(options.SinM,3));
            if license('test','Distrib_Computing_Toolbox')
                temp = options.SinM(:,:,:,uu);
                temp = [temp(:,end,:), temp, temp(:,1,:)];
                Ndist = options.Ndist;
                Nang = options.Nang;
                arc_interpolation = options.arc_interpolation;
                try
                    parfor kk = 1 : size(options.SinM,3)
                        if exist('scatteredInterpolant', 'file') == 3 && (~strcmp('cubic',arc_interpolation) && ~strcmp('v4',arc_interpolation))
                            F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double(temp(:,:,kk)), [],1));
                            uus_SinM(:,:,kk) = F(angle, distance);
                        else
                            uus_SinM(:,:,kk) = griddata(angle_o, distance_o, double(temp(:,:,kk)), angle, distance, arc_interpolation);
                        end
                    end
                catch
                    for kk = 1 : size(options.SinM,3)
                        if exist('scatteredInterpolant', 'file') == 3 && (~strcmp('cubic',arc_interpolation) && ~strcmp('v4',arc_interpolation))
                            F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double(temp(:,:,kk)), [],1));
                            uus_SinM(:,:,kk) = F(angle, distance);
                        else
                            uus_SinM(:,:,kk) = griddata(angle_o, distance_o, double(temp(:,:,kk)), angle, distance, arc_interpolation);
                        end
                    end
                end
            else
                for kk = 1 : size(options.SinM,3)
                    if exist('scatteredInterpolant', 'file') == 3 && (~strcmp('cubic',options.arc_interpolation) && ~strcmp('v4',options.arc_interpolation))
                        F = scatteredInterpolant(angle_o(:), distance_o(:), reshape(double([options.SinM(:,end,kk,uu), options.SinM(:,:,kk,uu), options.SinM(:,1,kk,uu)]), [],1));
                        uus_SinM(:,:,kk) = F(angle, distance);
                    else
                        uus_SinM(:,:,kk) = griddata(angle_o, distance_o, double([options.SinM(:,end,kk,uu), options.SinM(:,:,kk,uu), options.SinM(:,1,kk,uu)]), angle, distance, options.arc_interpolation);
                    end
                end
            end
            uus_SinM = uus_SinM(:,2:end-1,:);
            uus_SinM(isnan(uus_SinM)) = single(0);
            options.SinM(:,:,:,uu) = single(uus_SinM);
        end
        endTime = toc;
    end
    if options.verbose
        disp(['Arc correction complete in ' num2str(endTime) ' seconds'])
    end
end
end

File: source/Python/omegatomo/util/arcCorrection.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 25 11:40:44 2025

@author: Ville-Veikko Wettenhovi
"""

import numpy as np

def arc_correction(options, interpolate_sinogram):
    from omegatomo.projector.detcoord import detectorCoordinates, sinogramCoordinates2D
    xp, yp = detectorCoordinates(options)
    x_o, y_o = sinogramCoordinates2D(options, xp, yp)

    new_xp = np.zeros_like(xp)
    new_yp = np.zeros_like(yp)

    # Detector angles
    if options.blocks_per_ring % 4 == 0:
        l_angles = np.linspace(0, 90, options.blocks_per_ring // 4 + 1)
    else:
        l_angles = np.linspace(0, 180, options.blocks_per_ring // 2 + 1)
        l_angles = l_angles[:int(np.ceil(options.blocks_per_ring / 4))]
    ll = 0  # Python index

    # Shift to zero angle
    shift_val = int(options.offangle + (options.cryst_per_block + 1) / 2 if options.det_w_pseudo > options.det_per_ring else options.offangle + options.cryst_per_block / 2)
    xp = np.roll(xp, -shift_val)
    yp = np.roll(yp, -shift_val)

    # xp -= options.diameter / 2
    # yp -= options.diameter / 2

    quarter_len = len(xp) // 4
    for kk in range(quarter_len):
        if options.det_w_pseudo > options.det_per_ring:
            r = int(options.det_w_pseudo / 2)
            val_range = np.arange(kk + r - (options.cryst_per_block + 1), kk + r + (options.cryst_per_block + 1) + 1)
        else:
            r = int(options.det_w_pseudo / 2)
            val_range = np.arange(kk + r - options.cryst_per_block, kk + r + options.cryst_per_block + 1)

        angle_ref = l_angles[ll]
        dx = xp[kk] - xp[val_range]
        dy = yp[kk] - yp[val_range]
        angle1 = np.round(np.degrees(np.arctan2(dy, dx)) * 1e4) / 1e4

        if options.blocks_per_ring % 4 == 0:
            diffs = np.abs(np.abs(angle1) - angle_ref)
            ind1 = np.argmin(diffs)
        else:
            ind1 = np.argmin(np.abs(np.abs(angle1) - angle_ref))

        x2 = xp[val_range[ind1]]
        y2 = yp[val_range[ind1]]

        p = np.array([xp[kk], yp[kk]])
        q = np.array([x2, y2])
        d = q - p
        d_dot = np.dot(d, d)
        p_dot = np.dot(p, p)
        pd_dot = np.dot(2 * p, d)
        rad2 = (options.diameter / 2) ** 2

        sqrt_term = pd_dot ** 2 - 4 * d_dot * (p_dot - rad2)
        l = (-pd_dot - np.sqrt(sqrt_term)) / (2 * d_dot)

        lx = p + l * d
        new_xp[kk] = lx[0]
        new_yp[kk] = lx[1]

    # Mirror fill
    new_xp[:quarter_len] += options.diameter / 2
    new_yp[:quarter_len] += options.diameter / 2

    new_yp[quarter_len:2*quarter_len] = np.flip(new_yp[:quarter_len])
    diffi = np.diff(np.concatenate([np.flip(new_yp[:quarter_len]), [options.diameter/2]]))
    diffi = options.diameter/2 + np.cumsum(np.flip(diffi))
    new_yp[2*quarter_len:] = np.concatenate([diffi, np.flip(diffi)])

    diffi = np.diff(np.concatenate([new_xp[:quarter_len], [options.diameter/2]]))
    diffi = options.diameter/2 + np.cumsum(np.flip(diffi))
    new_xp[quarter_len:2*quarter_len] = diffi
    new_xp[2*quarter_len:] = np.flip(new_xp[:2*quarter_len])

    # Undo shift
    xp = np.roll(new_xp, shift_val)
    yp = np.roll(new_yp, shift_val)

    x, y = sinogramCoordinates2D(options, xp, yp)
    
    # Reshape LOR coordinates
    xx1 = x[:, 0].reshape(options.Ndist, options.Nang, order='F')
    xx2 = x[:, 1].reshape(options.Ndist, options.Nang, order='F')
    yy1 = y[:, 0].reshape(options.Ndist, options.Nang, order='F')
    yy2 = y[:, 1].reshape(options.Ndist, options.Nang, order='F')
    
    # Compute LOR angles (degrees), shift to avoid negatives, find J
    angle = np.degrees(np.arctan((yy1 - yy2) / (xx1 - xx2))) + 90
    angle[angle == 180] = 0
    J = np.argmin(np.mean(angle, axis=0))
    
    # Create arc-corrected coordinates from perpendicular LOR
    eka = xx1[0, J]
    vika = xx1[-1, J]
    alkux = np.linspace(eka, vika, options.Ndist)
    
    alkuy = np.sqrt((options.diameter / 2)**2 - (alkux - options.diameter / 2)**2) + options.diameter / 2
    alku = np.asfortranarray(np.vstack([alkux, alkuy]))
    alku2 = alku - options.diameter / 2
    
    alku = np.asfortranarray(np.vstack([alkux, np.abs(alkuy - options.diameter)]))
    alku1 = alku - options.diameter / 2
    
    # Compute angles
    angles = np.linspace(0, 180, options.Nang + 1)[:-1]
    angles = np.roll(angles, J + 1)
    angles = angles.reshape(1, 1, -1)
    
    # Rotation matrices
    cos_a = np.cos(np.radians(angles)).squeeze()
    sin_a = np.sin(np.radians(angles)).squeeze()
    rot_matrix = np.array([[cos_a, -sin_a], [sin_a, cos_a]], order='F')  # shape: (2, 2, Nang)
    
    # Rotate all points with each matrix (broadcasting)
    new_xy1 = np.zeros((options.Ndist * options.Nang,2),order='F')
    new_xy2 = np.zeros((options.Ndist * options.Nang,2),order='F')
    for ii in range(0, rot_matrix.shape[2]):
        new_xy1[ii * options.Ndist:(ii + 1) * options.Ndist] = (rot_matrix[:,:,ii] @ alku1).T + options.diameter / 2
        new_xy2[ii * options.Ndist:(ii + 1) * options.Ndist] = (rot_matrix[:,:,ii] @ alku2).T + options.diameter / 2
    # new_xy1 = np.einsum('ijk,k->ji', rot_matrix, alku1) + options.diameter / 2
    # new_xy2 = np.einsum('ijk,k->ji', rot_matrix, alku2) + options.diameter / 2

    # Multiply each matrix with alku1 and alku2 (2 x Ndist vectors)
    # Result will be shape (options.Nang, 2) after transpose
    # new_xy1 = np.array([(R @ alku1).T for R in rot_matrix]) + options.diameter / 2
    # new_xy2 = np.array([(R @ alku2).T for R in rot_matrix]) + options.diameter / 2
    # Update x and y arrays
    x = np.column_stack([new_xy1[:, 0], new_xy2[:, 0]])
    y = np.column_stack([new_xy1[:, 1], new_xy2[:, 1]])
    
    # Flip half of the LORs to match symmetry (MATLAB-style)
    xx1 = x[:, 0].reshape(options.Ndist, options.Nang, order='F')
    xx2 = x[:, 1].reshape(options.Ndist, options.Nang, order='F')
    yy1 = y[:, 0].reshape(options.Ndist, options.Nang, order='F')
    yy2 = y[:, 1].reshape(options.Ndist, options.Nang, order='F')
    
    apu = xx1[:, :J].copy()
    xx1[:, :J] = np.flipud(xx2[:, :J])
    xx2[:, :J] = np.flipud(apu)
    
    apu = yy1[:, :J].copy()
    yy1[:, :J] = np.flipud(yy2[:, :J])
    yy2[:, :J] = np.flipud(apu)
    
    # Flatten back to 2D coordinates
    x[:, 0] = xx1.ravel('F')
    x[:, 1] = xx2.ravel('F')
    y[:, 0] = yy1.ravel('F')
    y[:, 1] = yy2.ravel('F')

    x -= options.diameter / 2
    y -= options.diameter / 2
    
    # Reshape original coordinates too
    xx1_o = x_o[:, 0].reshape(options.Ndist, options.Nang, order='F')
    xx2_o = x_o[:, 1].reshape(options.Ndist, options.Nang, order='F')
    yy1_o = y_o[:, 0].reshape(options.Ndist, options.Nang, order='F')
    yy2_o = y_o[:, 1].reshape(options.Ndist, options.Nang, order='F')
    
    if interpolate_sinogram:

        from scipy.interpolate import griddata
        if options.SinM.size == options.SinM.shape[0] or options.SinM.ndim < 4:
            options.SinM = options.SinM.reshape((options.Ndist, options.Nang, options.NSinos, -1), order='F')
    
        angles_o = np.degrees(np.arctan((yy1_o - yy2_o) / (xx1_o - xx2_o)))
        angles_o -= np.min(angles_o[:, 0])
        angles_o[angles_o < 0] += 180
        angle_o = np.concatenate([
            -angles_o[:, [1]],
            angles_o,
            np.abs(angles_o[:, [0]] - 180)
        ], axis=1)
    
        angles = np.degrees(np.arctan((yy1 - yy2) / (xx1 - xx2)))
        angles -= np.min(angles[:, 0])
        angles[angles < 0] += 180
        angle = np.concatenate([
            -angles[:, [1]],
            angles,
            np.abs(angles[:, [0]] - 180)
        ], axis=1)
    
        # x0 = options.diameter / 2
        # y0 = options.diameter / 2
        x0 = 0.
        y0 = 0.
    
        def orth_dist(x1, y1, x2, y2):
            return np.abs((y1 - y2) * x0 - (x1 - x2) * y0 + x1 * y2 - y1 * x2) / \
                   np.sqrt((y1 - y2)**2 + (x1 - x2)**2)
    
        distance_o = orth_dist(x_o[:, 0], y_o[:, 0], x_o[:, 1], y_o[:, 1]).reshape(options.Ndist, options.Nang,order='F')
        distance_o[options.Ndist // 2:] *= -1
        distance_o = np.concatenate([
            distance_o[:, [0]],
            distance_o,
            distance_o[:, [0]]
        ], axis=1)
    
        distance = orth_dist(x[:, 0], y[:, 0], x[:, 1], y[:, 1]).reshape(options.Ndist, options.Nang, order='F')
        distance[options.Ndist // 2:] *= -1
        distance = np.concatenate([
            distance[:, [0]],
            distance,
            distance[:, [0]]
        ], axis=1)
    
        for uu in range(options.SinM.shape[3]):
            sinm = options.SinM[..., uu]
            sinm_ext = np.concatenate([
                sinm[:, -1:, :],
                sinm,
                sinm[:, :1, :]
            ], axis=1)
            
            for zz in range(sinm_ext.shape[2]):
                values = sinm_ext[:, :, zz].astype(np.float32)
            
                # Flatten the grid and values
                points = np.column_stack((angle_o.ravel('F'), distance_o.ravel('F')))
                vals = values.ravel('F')
            
                # Interpolate
                interpolated = griddata(
                    points,
                    vals,
                    (angle, distance),
                    method=options.arc_interpolation,
                    fill_value=0.0
                )
            
                interpolated[np.isnan(interpolated)] = 0.0
                options.SinM[:, :, zz, uu] = interpolated[:,1:interpolated.shape[1]-1].astype(np.float32)
    
        if options.verbose:
            print("Arc correction interpolation complete")

    return x, y, options